### ISSUE TITLE
[Feature request] A few things that numpy does that pytorch could have

### ISSUE THREAD

**Issue by martinarjovsky** [pytorch/pytorch#1875]
Hi! Small list of simple features that I used on a day to day basis both in numpy and theano and would be helpful to have now. They're all super dumb, but writing hack-arounds or debugging a few extra things is sometimes annoying, and they should be super easy to add.

- Add .mean() for ByteTensors so we can calculate accs easily.
```
x = torch.Tensor(3).normal_()
(x > 0).mean() # Throws error

y = np.random.randn(3)
(y > 0).mean() # Works ok
```

- It's weird that .random_() defaults to garbage and not U[0,1] or something like that. Seems very error prone (I myself had to debug this ~ 10 times). Maybe a warning at least? 
```
x = torch.Tensor(2,3)
x.random_() # Returns random ints from 0 to max_int?

y = np.random.random((2,3)) # Returns U[0,1] on each entry
```

- Having something like ones_like and zeros_like would be ideal! I used to use this all the time for masks and other things. Also the hack arounds are very prone to do a bad use of memory, as opposed to the expand_as [esp when the programmer is bad such as myself :) ].
```
y = np.random.rand(3)
z = np.ones_like(y)
```

- It's very annoying that torch.arange returns a FloatTensor and index_select requires a LongTensor. I guess when advanced indexing is in this won't matter :)

- Edit: Adding that np.arange works just by specifying the end (as range) and torch.arange doesn't. Also the numpy version has dtypo int64 while the torch one is a FloatTensor, though similar things like torch.randperm give LongTensor (this might be useful, idk).
```
y = np.arange(10) # [0, ..., 9]
x = torch.arange(10) # Hell unleashes
```

Cheers!

**Comment by aditya1702** [comment]
I want to start contributing. I have installed pytorch in develop mode and was intending to solve this issue. This may seem dumb but I dont know where do we add functions like max, min, arrange, ones_like, zeros_like. From the above pull request #2489 I saw that the new functions have been added to THCTensorMath.c. Do we need to write the code in C? or is there a python file where we add new functions?

**Comment by weiyangfb** [comment]
Sounds like most Numpy-like ops have already been implemented, except for the `torch.mean()` for ByteTensors. Any reason why it is not supported?

```python
def get_mean(dtype):
    try:
        data = np.random.randint(10, size=10)
        x = torch.tensor(data, dtype=dtype)
        print(x.dtype, x.mean())
    except Exception as e:
        print(e)
    
dtypes = [torch.float16, torch.float32, torch.float64, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]
for d in dtypes:
    get_mean(d)
```
mean is not implemented for type torch.HalfTensor
(torch.float32, tensor(5.5000))
(torch.float64, tensor(5.5000, dtype=torch.float64))
mean is not implemented for type torch.ByteTensor
mean is not implemented for type torch.CharTensor
mean is not implemented for type torch.ShortTensor
mean is not implemented for type torch.IntTensor
mean is not implemented for type torch.LongTensor

**Comment by gchanan** [comment]
CC @cpuhrsch.

It's not implemented because it was difficult to implement with the old torch libraries, but it should be feasible now.

**Comment by elvisyjlin** [comment]
The mean operation is still not implemented for several types as @weiyangfb mentions. However, in NumPy, those values are automatically converted to np.float64 when calculating their average values. Do you think it is better to have the same behavior of mean in PyTorch?

**Comment by app-tarush** [comment]
Can I take this to implement `torch.mean()` with same behaviour as in NumPy?

**Comment by mruberry** [comment]
> Hi! Small list of simple features that I used on a day to day basis both in numpy and theano and would be helpful to have now. They're all super dumb, but writing hack-arounds or debugging a few extra things is sometimes annoying, and they should be super easy to add.
> 
> * Add .mean() for ByteTensors so we can calculate accs easily.
> 
> ```
> x = torch.Tensor(3).normal_()
> (x > 0).mean() # Throws error
> 
> y = np.random.randn(3)
> (y > 0).mean() # Works ok
> ```

Good idea. Let's cover this with the more recent and general https://github.com/pytorch/pytorch/issues/45833.

> * It's weird that .random_() defaults to garbage and not U[0,1] or something like that. Seems very error prone (I myself had to debug this ~ 10 times). Maybe a warning at least? 

I filed https://github.com/pytorch/pytorch/issues/50332 to track this.

> * Having something like ones_like and zeros_like would be ideal! I used to use this all the time for masks and other things. Also the hack arounds are very prone to do a bad use of memory, as opposed to the expand_as [esp when the programmer is bad such as myself :) ].

Good news! We have [ones_like](https://pytorch.org/docs/master/generated/torch.ones_like.html?highlight=ones_like#torch.ones_like) and [zeros_like](https://pytorch.org/docs/master/generated/torch.zeros_like.html?highlight=zeros_like#torch.zeros_like) now.

> * It's very annoying that torch.arange returns a FloatTensor and index_select requires a LongTensor. I guess when advanced indexing is in this won't matter :)
> * Edit: Adding that np.arange works just by specifying the end (as range) and torch.arange doesn't. Also the numpy version has dtypo int64 while the torch one is a FloatTensor, though similar things like torch.randperm give LongTensor (this might be useful, idk).
> 
> ```
> y = np.arange(10) # [0, ..., 9]
> x = torch.arange(10) # Hell unleashes
> ```

More good news: [torch.arange](https://pytorch.org/docs/master/generated/torch.arange.html?highlight=arange#torch.arange) now works as expected.

I'm going to close this issue because it's a multipart issue and several of its parts appear resolved. The two more modern issues are more focused and hence more actionable.
